cuqi/testproblem/_testproblem.py

```diff
@@ -254,6 +297,94 @@ def __init__(self,
         self.exactData = y_exact
         self.infoString = "Noise type: Additive {} with std: {}".format(noise_type.capitalize(),noise_std)
 
+def _getConvolutionOperator(dim, PSF, PSF_param, PSF_size, BC):
+
+    # Boundary condition translation
+    if BC.lower() == "zero":
+        mode = "constant"
+    elif BC.lower() == "periodic":
+        mode = "wrap"
+    elif BC.lower() == "Neumann":
+        mode = "symmetric"
+    elif BC.lower() == "Mirror":
+        mode = "reflect"
+    elif BC.lower() == "Nearest":
+        mode = "edge"
+    else:
+        raise ValueError("Unknown boundary condition")
+
+    if PSF_size is None:
+        PSF_size = dim
+   
+    # PSF setup
+    if isinstance(PSF, np.ndarray):
+        if PSF.ndim != 1:
+            raise ValueError("PSF must be a 1D array")
+        P = PSF
+    elif isinstance(PSF, str):
+        if PSF.lower() == "gauss":
+            P, _ = _GaussPSF_1D(PSF_size, PSF_param)
+        elif PSF.lower() == "moffat":
+            P, _ = _MoffatPSF_1D(PSF_size, PSF_param)
+        elif PSF.lower() == "defocus":
+            P, _ = _DefocusPSF_1D(PSF_size, PSF_param)
+        else:
+            raise ValueError("Unknown PSF type")
+    
+    # Convolution matrix
+    return lambda x: convolve1d(x, P, mode=mode)
+
+
+def _GaussPSF_1D(PSF_size, PSF_param):
+    if PSF_param is None:
+        PSF_param = 10
+
+    # Set up grid points to evaluate the Gaussian function
+    x = np.arange(-np.fix(PSF_size/2), np.ceil(PSF_size/2))
+
+    # Compute the Gaussian, and normalize the PSF.
+    PSF = np.exp( -0.5*((x**2)/(PSF_param**2)) )
+    PSF /= PSF.sum()
+
+    # find the center
+    center = np.where(PSF == PSF.max())[0][0]
+    return PSF, center.astype(int)
+
+def _MoffatPSF_1D(PSF_size, PSF_param, beta=1):
```

Comment:
_MoffatPSF_1D and _GaussPSF_1D and to some extent _DefocusPSF_1D have repeated code lines, could be refactored.

Reply:
Good find